Source/render.cpp

```diff
@@ -422,306 +539,834 @@ inline void DoRenderLine(BYTE *dst, BYTE *src, size_t n, BYTE *tbl, DWORD mask)
+	for (auto i = 1 + clipY.upperBottom; i <= upperMax; ++i, dst -= dstPitch - XStep) {
+		memset(dst, 0, TILE_WIDTH - 2 * XStep * i);
 	}
 }
 
 // Blit a black tile without clipping (must be fully in bounds).
-void BlitBlackTileFull(BYTE *dst, int dst_pitch)
+void RenderBlackTileFull(BYTE *dst, int dstPitch)
 {
+	dst += XStep * (LowerHeight - 1);
 	// Tile is fully in bounds, can use constant loop boundaries.
 	// Lower triangle (drawn bottom to top):
-	for (int i = 1; i <= LowerHeight; ++i, dst -= dst_pitch + XStep) {
+	for (unsigned i = 1; i <= LowerHeight; ++i, dst -= dstPitch + XStep) {
 		memset(dst, 0, 2 * XStep * i);
 	}
 	dst += 2 * XStep;
 	// Upper triangle (drawn bottom to to top):
-	for (int i = 0, j = UpperHeight; i < UpperHeight; ++i, --j, dst -= dst_pitch - XStep) {
-		memset(dst, 0, 2 * XStep * j);
+	for (unsigned i = 1; i <= TriangleUpperHeight; ++i, dst -= dstPitch - XStep) {
+		memset(dst, 0, TILE_WIDTH - 2 * XStep * i);
 	}
 }
 
 } // namespace
 
-void world_draw_black_tile(const CelOutputBuffer &out, int sx, int sy)
+void RenderTile(const CelOutputBuffer &out, int x, int y)
 {
-	if (sx <= -TILE_WIDTH || sx >= out.region.w || sy < 0 || sy + 1 > out.region.h + Height)
+	const auto tile = static_cast<TileType>((level_cel_block & 0x7000) >> 12);
+	const auto *mask = GetMask(tile);
+	if (mask == nullptr)
 		return;
 
-	// Initial out position: the bottom corner of the lower triangle.
-	int out_x = sx + TILE_WIDTH / 2 - XStep;
-	int out_y = sy;
-
-	// How many lines to skip for the lower and upper triangles:
-	int skip_lower_bottom = 0;
-	int skip_lower_top = 0;
-	int skip_upper_bottom = 0;
-	int skip_upper_top = 0;
-	if (sy >= out.region.h) {
-		skip_lower_bottom = sy - out.region.h + 1;
-		out_y -= skip_lower_bottom;
-		if (skip_lower_bottom > LowerHeight) {
-			skip_upper_bottom = skip_lower_bottom - LowerHeight;
-			skip_lower_bottom = LowerHeight;
-		}
-		out_x += XStep * (skip_upper_bottom - skip_lower_bottom);
-	} else if (sy + 1 < Height) {
-		skip_upper_top = Height - (sy + 1) - 1;
-		if (skip_upper_top > UpperHeight) {
-			skip_lower_top = skip_upper_top - UpperHeight;
-			skip_upper_top = UpperHeight;
-		}
+#ifdef DEBUG_RENDER_OFFSET_X
+	x += DEBUG_RENDER_OFFSET_X;
+#endif
+#ifdef DEBUG_RENDER_OFFSET_Y
+	y += DEBUG_RENDER_OFFSET_Y;
+#endif
+#ifdef DEBUG_RENDER_COLOR
+	DBGCOLOR = GetTileDebugColor(tile);
+#endif
+
+	Clip clip = CalculateClip(x, y, Width, GetTileHeight(tile), out);
+	if (clip.width <= 0 || clip.height <= 0)
+		return;
+
+	const BYTE *tbl = &pLightTbl[256 * light_table_index];
+	const auto *pFrameTable = reinterpret_cast<const std::uint32_t *>(pDungeonCels.get());
+	BYTE *src = &pDungeonCels[SDL_SwapLE32(pFrameTable[level_cel_block & 0xFFF])];
+	BYTE *dst = out.at(static_cast<int>(x + clip.left), static_cast<int>(y - clip.bottom));
+	const auto dstPitch = out.pitch();
+	mask -= clip.bottom;
+
+	switch (tile) {
+	case TileType::Square:
+		RenderSquare(dst, dstPitch, src, mask, tbl, clip);

```

Comment:
Didn't you just rename my "square" map function to diamond :)

Reply:
Your "square" is a diamond, this is an actual square. See the TileType enum documentation

Reply:
ok :)